### PR TITLE
Do not recurse when the service address is a local cname

### LINKED ISSUE #2795: Consul Server Crashing on DNS lookup
### `consul version` for both Client and Server

```
Consul v0.7.5
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```

### `consul info` for both Client and Server

Server
```
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 3
build:
        prerelease = 
        revision = '21f2d5a
        version = 0.7.5
consul:
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = 192.168.23.101:8300
        server = true
raft:
        applied_index = 511360
        commit_index = 511360
        fsm_pending = 0
        last_contact = 65.233681ms
        last_log_index = 511360
        last_log_term = 119
        last_snapshot_index = 507928
        last_snapshot_term = 106
        latest_configuration = [{Suffrage:Voter ID:192.168.23.100:8300 Address:192.168.23.100:8300} {Suffrage:Voter ID:192.168.23.102:8300 Address:192.168.23.102:8300} {Suffrage:Voter ID:192.168.23.101:8300 Address:192.168.23.101:8300}]
        latest_configuration_index = 1
        num_peers = 2
        protocol_version = 1
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 119
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 98
        max_procs = 2
        os = linux
        version = go1.7.5
serf_lan:
        encrypted = true
        event_queue = 0
        event_time = 30
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 479
        members = 22
        query_queue = 0
        query_time = 1
serf_wan:
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1
```

Client
```
agent:
        check_monitors = 0
        check_ttls = 1
        checks = 1
        services = 2
build:
        prerelease = 
        revision = '21f2d5a
        version = 0.7.5
consul:
        known_servers = 3
        server = false
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 43
        max_procs = 2
        os = linux
        version = go1.7.5
serf_lan:
        encrypted = true
        event_queue = 0
        event_time = 30
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 479
        members = 22
        query_queue = 0
        query_time = 1
```

### Operating system and Environment details

Centos 7 with 3 Consul Servers and ~20 consul clients

### Description of the Issue (and unexpected/desired result)

I added Vault to the Consul cluster configured to register it's own service.

When I lookup the service via DNS the Consul Server will randomly crash with the huge stack trace linked bellow and well as giving the following error when using Dig

```
[root@consul-server1 ~]# dig @127.0.0.1 -p 8600 vault.service.consul.rmb938.me 

; <<>> DiG 9.9.4-RedHat-9.9.4-38.el7_3.2 <<>> @127.0.0.1 -p 8600 vault.service.consul.rmb938.me
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
```

Querying any other service DNS record works fine without any issues. This only ever seems to happen when trying to lookup vault.

### Log Fragments or Link to [gist](https://gist.github.com/)

https://gist.github.com/rmb938/33f836c69c35a4373926cc386325dd3f

### PR DESCRIPTION
**Pull request by magiconair**
Fixes #2795

### CODE REVIEW

**Comment by magiconair** [comment]
This fixes #2795 and doesn't break any test but I would need some guidance whether this is the right approach. Also, I still need to add a test for this.

File: command/agent/dns.go [hunk 1]
```diff
@@ -833,6 +833,12 @@ func (d *DNSServer) serviceNodeRecords(dc string, nodes structs.CheckServiceNode
 			addr = node.Service.Address
 		}
 
+		// If the service address is a CNAME for the service we are looking
+		// for then use the node address.
+		if qName == addr+"." {
```
Comment (slackpad) [inline command/agent/dns.go]: This should work if they put the dot at the end of the addr as well:

```
if qName == strings.TrimSuffix(".")+"."
```

**Comment by slackpad** [comment]
This looks good if we add a unit test for it.

File: command/agent/dns.go (hunk 1 above)
Comment (magiconair) [inline command/agent/dns.go]: I'll try that and write a test.

**Comment by magiconair** [comment]
@slackpad added a test using `verify.Values` Pls let me know if the style is OK. I can convert other tests to that format in a subsequent PR.

**Pull request merged**